extensions/amp-viewer-integration/0.1/amp-viewer-integration.js

```diff
@@ -14,5 +14,73 @@
  * limitations under the License.
  */
 
-export class AmpViewerIntegration {
+import {Messaging} from './messaging.js';
+import {listen, unlisten} from '../../../src/event-helper';
+import {viewerPromiseForDoc} from '../../../src/viewer';
+import {user} from '../../../src/log';
+
+
+class AmpViewerIntegration {
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    /** @const @private {!Window} win */
+    this.win_ = win;
+  }
+
+  init() {
+    let viewer;
+    this.getViewer()
+    .then(viewerParam => {
+      viewer = viewerParam;
+      return this.getHandshakePromise(viewer);
+    })
+    .then(viewerOrigin => {
+      const messaging = new Messaging(this.win_.parent, viewerOrigin,
+          function(type, payload, awaitResponse) {
+            return viewer.receiveMessage(type, payload, awaitResponse);
+          }, this.win_.location.href);
+      viewer.setMessageDeliverer(function(type, payload, awaitResponse) {
+        return messaging.sendRequest(type, payload, awaitResponse);
+      }, viewerOrigin);
+    });
+  }
+
+  /**
+   * @return {!Promise<!../../../src/service/viewer-impl.Viewer>}
+   */
+  getViewer() {
+    return viewerPromiseForDoc(this.win_.document);
```

Comment:
Viewer is guaranteed to exist in the extension. So feel free to just resolve it via `viewerFor` and you can even do this in the constructor if desired. Though, you reasoning for doing a promise here could also be valid. Let me know.